test/unit/pool/test-destroy-connection.js

```diff
@@ -11,10 +11,10 @@ server.listen(common.fakeServerPort, function (err) {
   pool.getConnection(function (err, connection) {
     assert.ifError(err);
 
-    assert.strictEqual(connection, pool._allConnections[0]);
+    assert.strictEqual(connection, pool._connectionManager._allConnections._map[1]);
```

Comment:
@ifsnow I think a lot of the pool optimizations are worth it, and certainly can make it into a major version of this module. The only real concern I have is that because this is just one giant PR, making good (but breaking) changes _and_ making good but non-breaking changes, together, I have no option but to delay all the features in this PR until the next major. If you can simply break it up into the different pieces as multiple PRs, we can easily accept the easy ones now and discuss the breaking ones better and decide on the release of a new major version.